src/main/java/uk/gov/hmcts/reform/divorce/orchestration/config/courtallocation/CourtAllocationConfigurator.java

```
@@ -0,0 +1,105 @@
+package uk.gov.hmcts.reform.divorce.orchestration.config.courtallocation;
+
+import com.jayway.jsonpath.DocumentContext;
+import com.jayway.jsonpath.JsonPath;
+import com.jayway.jsonpath.Option;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import uk.gov.hmcts.reform.divorce.orchestration.courtallocation.CourtAllocationConfiguration;
+import uk.gov.hmcts.reform.divorce.orchestration.courtallocation.CourtAllocator;
+import uk.gov.hmcts.reform.divorce.orchestration.courtallocation.DefaultCourtAllocator;
+
+import java.math.BigDecimal;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toMap;
+
+@Configuration
+public class CourtAllocationConfigurator {
+
+    @Autowired
+    private CourtAllocationConfiguration courtAllocationConfig;
+
+    @Bean
+    public CourtAllocator configureCourtAllocationFromEnvironmentVariable() {
+        return new DefaultCourtAllocator(courtAllocationConfig.getDesiredWorkloadPerCourt(),
+            courtAllocationConfig.getDivorceRatioPerFact(),
+            courtAllocationConfig.getSpecificCourtsAllocationPerFact());
+    }
+
+    @Bean
+    public CourtAllocationConfiguration setUpCourtAllocationConfiguration(
+        @Value("${courtAllocationConfigurationJson}") String courtAllocationConfigJson) {
+        DocumentContext parsedJson = parseJsonConfiguration(courtAllocationConfigJson);
+
+        Map<String, BigDecimal> desiredWorkloadPerCourt = prepareDesiredWorkloadPerCourt(parsedJson);
+
+        Map<String, BigDecimal> divorceRatioPerFact = prepareDivorceRatioPerFact();
+
+        Map<String, Map<String, BigDecimal>> specificCourtsAllocationPerFact =
```

Question - do we validate the distribution? I.e. do we check it all adds up to 100%, no more, or less, within a set margin? This would validate the input config.

---

We get a few validations for free. For instance things like court and fact are not duplicated (since they get transformed into Maps in which they are keys).
We check that allocation doesn't go over 100% and that it's not 0%. We also check that a court is not overallocated.
We don't have a specific error for when the court allocation is under 100%, however, I've written an exploratory test to make sure that the probability library we use deals with having less than 100% assigned. I believe it handles it proportionately.

Do you think we should throw an error is the courts allocation don't match 100% exactly?

---

I was more thinking about if the person updating the config made a mistake, we could catch early by making sure their inputs added up. 

It's good what you say in terms of validation down-stream, but that is assuming the configurator has left a good set of numbers, and they add up correctly. Basically it's a sanity check on the config being used. Is this covered as well or...?

---

Cool. I'll implement it to be at least 99% allocation

---

Actually, I'll do 100%. The user can do 0.34 and 0.66 for distributing 1/3 and 2/3.

---

Done!